tide_site_restriction.module

```diff
@@ -162,3 +163,75 @@ function tide_site_restriction_form_views_exposed_form_alter(&$form, FormStateIn
     }
   }
 }
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function tide_site_restriction_form_user_form_alter(&$form, FormStateInterface $form_state, $form_id) {
+  if (isset($form['field_user_site'])) {
+    $roles = Role::loadMultiple();
+    $states = [];
+    foreach ($roles as $role_id => $role) {
+      if ($role->hasPermission('bypass node access') || $role->hasPermission('administer nodes') || $role->hasPermission('bypass site restriction')) {
+        $states[] = [
+          [':input[name="roles[' . $role_id . ']"]' => ['checked' => TRUE]],
+          'or',
+          // We have drupal/role_delegation dependency, so need to check
+          // role_change name.
+          [':input[name="role_change[' . $role_id . ']"]' => ['checked' => TRUE]],
+          'or',
+        ];
+      }
+    }
+    $results = [];
+    if ($states) {
+      foreach ($states as $state) {
+        $results = array_merge($results, $state);
+      }
+      // Removes the last 'or'.
+      if (end($results) == 'or') {
+        array_pop($results);
+      }
+      $form['field_user_site']['widget']['#states'] = [
+        'invisible' => $results,
+      ];
+    }
+    $form['#validate'][] = '_tide_site_restriction_validation';
+  }
+}
+
+/**
+ * Custom validation.
+ */
+function _tide_site_restriction_validation(&$form, FormStateInterface $form_state) {
+  if (!isset($form['role_change']['#access']) || !$form['role_change']['#access']) {
+    $values = $form_state->getValue('roles');
+  }
+  else {
+    $values = array_column($form_state->getValue('role_change'), 'target_id');
+  }
```

Comment:
Unless there's a reason not to, I'd re-factor this and nest the setting of the `$values` var inside a check of the status of role_delegation i.e. 
```suggestion
  if (\Drupal::service('module_handler')->moduleExists(['role_delegation']) {
    $values = array_column($form_state->getValue('role_change'), 'target_id');
   } else {
    $values = $form_state->getValue('roles');
   }
```
```